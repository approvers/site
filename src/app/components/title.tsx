import { Heading } from "@chakra-ui/react";
import type { ReactNode } from "react";

export const Title = ({ children }: { children?: ReactNode }): JSX.Element => (
  <Heading as="h1" w="100%" mt={8} mb={8} fontSize="2xl" textAlign="center">
    {children}
  </Heading>
);

export const Subtitle = ({ children }: { children?: ReactNode }): JSX.Element => (
  <Heading fontSize="xl" lineHeight="xl">
    {children}
  </Heading>
);
