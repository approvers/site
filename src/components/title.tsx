import type { ReactNode } from "react";
import styles from "../scss/components/title.module.scss";

export const Title = ({ children }: { children?: ReactNode }): JSX.Element => (
  <h1 className={styles.title}>{children}</h1>
);

export const Subtitle = ({ children }: { children?: ReactNode }): JSX.Element => (
  <h2 className={styles.subtitle}>{children}</h2>
);
