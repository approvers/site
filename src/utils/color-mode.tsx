"use client";

import type { IconButtonProps, SpanProps } from "@chakra-ui/react";
import { ClientOnly, IconButton, Skeleton, Span } from "@chakra-ui/react";
import { ThemeProvider, useTheme } from "next-themes";
import type { ThemeProviderProps } from "next-themes";
import * as React from "react";
import { LuMoon, LuSun } from "react-icons/lu";

export interface ColorModeProviderProps extends ThemeProviderProps {}

export function ColorModeProvider(props: ColorModeProviderProps) {
  return <ThemeProvider attribute="class" disableTransitionOnChange {...props} />;
}

export type ColorMode = "light" | "dark";

export interface UseColorModeReturn {
  colorMode: ColorMode;
  setColorMode: (colorMode: ColorMode) => void;
  toggleColorMode: () => void;
}

export function useColorMode(): UseColorModeReturn {
  const { resolvedTheme, setTheme, forcedTheme } = useTheme();
  const colorMode = forcedTheme || resolvedTheme;
  const toggleColorMode = () => {
    setTheme(resolvedTheme === "dark" ? "light" : "dark");
  };
  return {
    colorMode: colorMode as ColorMode,
    setColorMode: setTheme,
    toggleColorMode,
  };
}

export function useColorModeValue<T>(light: T, dark: T) {
  const { colorMode } = useColorMode();
  return colorMode === "dark" ? dark : light;
}

export function ColorModeIcon() {
  const { colorMode } = useColorMode();
  return colorMode === "dark" ? <LuMoon /> : <LuSun />;
}

export function OppositeColorModeIcon() {
  const { colorMode } = useColorMode();
  return colorMode === "dark" ? <LuSun /> : <LuMoon />;
}

interface ColorModeButtonProps extends Omit<IconButtonProps, "aria-label"> {}

export const ColorModeButton = React.forwardRef<HTMLButtonElement, ColorModeButtonProps>(
  function ColorModeButton(props, ref) {
    const { toggleColorMode } = useColorMode();
    return (
      <ClientOnly fallback={<Skeleton boxSize="9" />}>
        <IconButton
          ref={ref}
          aria-label="Toggle color mode"
          onClick={toggleColorMode}
          size="sm"
          variant="ghost"
          {...props}
          css={{
            _icon: {
              width: "5",
              height: "5",
            },
          }}
        >
          <ColorModeIcon />
        </IconButton>
      </ClientOnly>
    );
  },
);

export const LightMode = React.forwardRef<HTMLSpanElement, SpanProps>(
  function LightMode(props, ref) {
    return (
      <Span
        className="chakra-theme light"
        ref={ref}
        display="contents"
        color="fg"
        colorPalette="gray"
        colorScheme="light"
        {...props}
      />
    );
  },
);

export const DarkMode = React.forwardRef<HTMLSpanElement, SpanProps>(function DarkMode(props, ref) {
  return (
    <Span
      className="chakra-theme dark"
      ref={ref}
      display="contents"
      color="fg"
      colorPalette="gray"
      colorScheme="dark"
      {...props}
    />
  );
});
