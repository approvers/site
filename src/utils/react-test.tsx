import reactTestRenderer, { ReactTestRenderer, TestRendererOptions } from "react-test-renderer";
import { ReactElement } from "react";

export const render = (
  ui: ReactElement,
  options?: TestRendererOptions | undefined,
): ReactTestRenderer => reactTestRenderer.create(ui, options);
