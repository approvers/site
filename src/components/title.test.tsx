import { expect, it } from "vitest";

import { render } from "../utils/react-test";
import { Subtitle, Title } from "./title";

it("renders title correctly", () => {
  const tree = render(<Title>Epic Title</Title>);
  expect(tree.toJSON()).toMatchSnapshot();
});

it("renders subtitle correctly", () => {
  const tree = render(<Subtitle>Epic Subtitle</Subtitle>);
  expect(tree.toJSON()).toMatchSnapshot();
});
