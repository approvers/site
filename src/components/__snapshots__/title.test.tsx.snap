// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders subtitle correctly 1`] = `
<h2
  className="chakra-heading css-a29miu"
>
  Epic Subtitle
</h2>
`;

exports[`renders title correctly 1`] = `
<div
  className="css-x8fcf4"
>
  <h1
    className="chakra-heading css-1n49pl7"
  >
    Epic Title
  </h1>
</div>
`;
